presto-main/src/test/java/io/prestosql/sql/query/TestSubqueries.java

```diff
@@ -119,8 +143,113 @@ public void testCorrelatedSubqueriesWithLimit()
+        assertions.assertQuery(
+                "SELECT * " +
+                        "FROM (VALUES 1, 2, 3, null) outer_relation(id) " +
+                        "LEFT JOIN LATERAL " +
+                        "(SELECT value FROM " +
+                        "(VALUES " +
+                        "(1, 'd'), " +
+                        "(1, 'c'), " +
+                        "(1, 'b'), " +
+                        "(1, 'a'), " +
+                        "(2, 'w'), " +
+                        "(null, 'x')) inner_relation(id, value) " +
+                        "WHERE outer_relation.id = inner_relation.id " +
+                        "ORDER BY inner_relation.value LIMIT 2) " +
+                        "ON TRUE",
+                "VALUES " +
+                        "(1, 'a'), " +
+                        "(1, 'b'), " +
+                        "(2, 'w'), " +
+                        "(3, null), " +
+                        "(null, null)");
+        // correlated symbol in predicate not bound to inner relation + Limit
+        assertions.assertQuery(
+                "SELECT * " +
+                        "FROM (VALUES 1, 2, 3, null) outer_relation(id) " +
+                        "LEFT JOIN LATERAL " +
+                        "(SELECT value FROM (VALUES 'a', 'a', 'a') inner_relation(value) " +
+                        "   WHERE outer_relation.id = 3 LIMIT 2) " +
+                        "ON TRUE",
+                "VALUES (1, null), (2, null), (3, 'a'), (3, 'a'), (null, null)");
+        assertions.assertQuery(
+                "SELECT * " +
+                        "FROM (VALUES 1, 2, 3, null) outer_relation(id) " +
+                        "LEFT JOIN LATERAL " +
+                        "(SELECT 1 FROM (VALUES 'a', 'a', 'a') inner_relation(value) " +
+                        "   WHERE outer_relation.id = 3 LIMIT 2) " +
+                        "ON TRUE",
+                "VALUES (1, null), (2, null), (3, 1), (3, 1), (null, null)");
+        // correlated symbol in predicate not bound to inner relation + TopN
+        assertions.assertQuery(
+                "SELECT * " +
+                        "FROM (VALUES 1, 2, 3, null) outer_relation(id) " +
+                        "LEFT JOIN LATERAL " +
+                        "(SELECT value FROM (VALUES 'c', 'a', 'b') inner_relation(value) " +
+                        "   WHERE outer_relation.id = 3 ORDER BY value LIMIT 2) " +
+                        "ON TRUE",
+                "VALUES (1, null), (2, null), (3, 'a'), (3, 'b'), (null, null)");
+        // TopN with ordering not decorrelating
+        assertions.assertFails(
+                "SELECT * " +
+                        "FROM (VALUES 1, 2, 3, null) outer_relation(id) " +
+                        "LEFT JOIN LATERAL " +
+                        "(SELECT value FROM (VALUES 'c', 'a', 'b') inner_relation(value) " +
+                        "   WHERE outer_relation.id = 3 ORDER BY outer_relation.id LIMIT 2) " +
```

Comment:
Shouldn't `decorrelateOrderingScheme` decorrelate `outer_relation.id`?

Reply:
The correlation symbol in `ORDER BY` is resolved; the query fails due to projection `outer_relation.id -> outer_relation.id` not decorrelating.

Reply:
I wonder why this projection appears on the first place. it's not relevant though.